hack/generated/pkg/reconcilers/azure_deployment_reconciler.go

```diff
@@ -0,0 +1,869 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package reconcilers
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	autorestAzure "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/k8s-infra/pkg/util/ownerutil"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/Azure/k8s-infra/hack/generated/pkg/armclient"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/genruntime"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/reflecthelpers"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/util/armresourceresolver"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/util/kubeclient"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/util/patch"
+)
+
+// TODO: I think we will want to pull some of this back into the Generic Controller so that it happens
+// TODO: for all resources
+
+const (
+	// TODO: Delete these later in favor of something in status?
+	DeploymentIdAnnotation   = "deployment-id.infra.azure.com"
+	DeploymentNameAnnotation = "deployment-name.infra.azure.com"
+	ResourceStateAnnotation  = "resource-state.infra.azure.com"
+	ResourceIdAnnotation     = "resource-id.infra.azure.com"
+	ResourceErrorAnnotation  = "resource-error.infra.azure.com"
+	ResourceSigAnnotationKey = "resource-sig.infra.azure.com"
+	// PreserveDeploymentAnnotation is the key which tells the applier to keep or delete the deployment
+	PreserveDeploymentAnnotation = "x-preserve-deployment"
+)
+
+// TODO: Do we actually want this at the controller level or this level?
+const GenericControllerFinalizer = "generated.infra.azure.com/finalizer"
+
+type CreateOrUpdateAction string
+
+const (
+	CreateOrUpdateActionNoAction          = CreateOrUpdateAction("NoAction")
+	CreateOrUpdateActionManageOwnership   = CreateOrUpdateAction("ManageOwnership")
+	CreateOrUpdateActionBeginDeployment   = CreateOrUpdateAction("BeginDeployment")
+	CreateOrUpdateActionMonitorDeployment = CreateOrUpdateAction("MonitorDeployment")
```

Comment:
Given that not everything is "deployed" (e.g. database users, network permissions, etc) perhaps these should be "BeginUpdate" and "MonitorUpdate" ?

Reply:
This is the `deployment_reconciler` though so in this context I think it's ok/correct. The intent was for other things (created through something that isn't the Azure `deployment` ARM infrastructure) to have their own reconcilers, which is why I changed the `generic_controller` to operate at the `reconciler` level and abstracted all of this deployment specific stuff out.